### ISSUE TITLE
Malicious actors can programmatically get full points

### ISSUE THREAD

**Issue by 18chiatt** [Carnap/Carnap#224]
REDACTED by @gleachkr - Sorry, I realize that security by obscurity is bad, but I don't want to tempt students who are casually reading through, and who might copy-paste before they get the end of the discussion below, where they are warned that this trick is detectable after the fact and might make them fail their class if they try it.

**Comment by 18chiatt** [comment]
Additionally, no attempt is made to check if the problem submitted for is actually a real problem, even if the problem does not exist, you will be granted 5 points by default, allowing for nearly unlimited points

**Comment by JakeEhrlich** [comment]
Graham would know better but I think early on we decided that if I student could do this we wouldn’t bother putting anything in their way to stop them. I think there’s even easier ways to so it from the browser by manipulating DOM or JavaScript. To be fair though I don’t think we imagined this many students or covid at the time. Especially for tests the ethical thing to do is probably to recheck server side.

**Comment by JakeEhrlich** [comment]
Does each course have its own sever? There may be performance issues (im so far out of there loop nowadays I don’t now how long proof checking typically takes). Certainly if Carnap scaled even further we’d eventually hit that.

**Comment by gleachkr** [comment]
Yeah, I think that's what we were thinking---the intersection between the set of students who can do this and the set of students would want to was pretty small. But now that the project is bigger, that set is bigger.

The reason for not checking on the server side is partly an efficiency issue. Proof checking is expensive, and one thing that's nice about the current design is that it lets us scale well because the more users we have, the more CPUs we can proof check with.

> Does each course have its own sever? 

Right now there's just one. Down the road, it's looking like it'll probably be per-university, or possibly like, one university might share with others nearby, e.g. something like one server for the Cal state system perhaps.

But: the thing about this hack (clever, btw, nice job @18chiatt) is that you could be detected after the fact, and the consequences for you being detected would be pretty bad. This would already show up as anomolous if the instructor reviews the assignment for partial credit. But there might be ways to flag this kind of anomolous thing for instructors more effectively. Minimally, raise an explicit flag in the review area when there's anything like this associated with the assignment. Maybe ideally, have a single button "check integrity" action on the server. The after-the-fact checks could all run on the instructor side, just like the checks in the review area do, so you wouldn't need to worry about slamming the server.

**Comment by 18chiatt** [comment]
The Javascript is all minified/obfuscated, making it quite difficult to directly attack through that.  I would recommend adding some kind of "signature" which in some way includes the time, the question content, and maybe something about the correct answer when submitting answers.  This fingerprint could use some strange combination of these things which would be difficult to calculate outside the browser or reverse engineer, while being trivial to verify on the server side.

**Comment by JakeEhrlich** [comment]
Yeah integrity check in the instructors browser and maybe even some kind of anomaly detector to spot anything we don’t expect sounds like it would scale well and would essentially eliminate any realistic attempt at getting away with this.

**Comment by rzach** [comment]
If someone were paranoid that this was happening, how would they go about checking for it "after the fact"?

**Comment by gleachkr** [comment]
You could go to the review area and make sure that the student 
submission actually had a proof attached to it. Of course, again, it 
would be better if it were possible to automate this kind of check.

Another idea that I was mulling was to have Carnap's submission function 
also incorporate some kind of hash of the problem contents, which would 
then be verified on the server side. This wouldn't be foolproof, but it 
would make it considerably harder to send in fake submissions, since 
you'd have to actually understand how the submission hashing works.

**Comment by lf-** [comment]
> The Javascript is all minified/obfuscated, making it quite difficult to directly attack through that. I would recommend including some kind of "signature" in the request which in some way includes the time, the question content, and maybe something about the correct answer when submitting answers. This fingerprint could use some strange combination of these things which would be difficult to calculate outside the browser or reverse engineer, while being trivial to verify on the server side.
> 
> especially with this fingerprint addition, it would be very obvious if somebody used a user-generated http request

This is ultimately security by obscurity, which would not work well against a particularly determined attacker with access to the open source source code (and could thus maybe generate a source map or other similar thing and make using a javascript debugger to break it more feasible).

I know our proof checker is unminified on account of it being 42MB (yeah...) which compresses down to 5MB. It is on our backlog to figure out why this is the case and reduce the size. Regardless, it's compiler output from a dissimilar language, and Haskell compiler output at that so it's unlikely anyone would *want* to touch it. I don't think any of the Carnap site JS bits are minified either.

I like the proposal to run the proof checker on a batch of student answers on an instructor's computer overnight or similar, if it is a significant computational issue. We could alternatively build a server side system that could run such batches in the background for instance.

Something else we could do is disable the client side checker entirely for certain high stakes assignments for example. That would discourage anyone messing with them since the field to tamper with wouldn't be there at all and instructors would not get a potential false sense of security before checking the proofs on a trusted computer.

**Comment by rzach** [comment]
Was this supposed to be closed?

**Comment by gleachkr** [comment]
No, unfortunately. Typo in a commit message meant to close #244.
